### PR TITLE
799 LB refactor and cleanup

### LINKED ISSUE #799: Refactor LB using proper components in support of pre-LB stats
**What Needs to be Done?**

- [x] See Section 3.1 in the report SAND2020-3967.

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #799 

- Major refactor of LB system.
  - `ProcStats` was completely static, which is a problem. It needs to be contained in an instance, now a new objgroup component `theProcStats` with proper dependencies
  - `StatsRestartReader` was piggybacking off of `ProcStats` in an improper manner; it's quite separate and shouldn't be pushing data through `ProcStats`. Refactor it as a component itself

All this refactor enables the `LBManager` to talk to `ProcStats` (now a proper component) when making a decision on the LB to pick and when to run it.

### CODE REVIEW

File: src/vt/vrt/collection/balance/lb_invoke/invoke.cc [hunk 1]
```diff
@@ -135,35 +141,61 @@ void LBManager::collectiveImpl(
...
+}
+
+void LBManager::computedCoeffVariance(Coeff eff) {
+  // Always return on 0'th phase---no coefficient of variation
+  if (phase_impl_ < 1) {
+    return releaseNow(phase_impl_);
+  }
+
+  auto const coeff_mean = eff.mean();
+  auto const coeff_max = eff.max();
```
Comment (lifflander) [inline src/vt/vrt/collection/balance/lb_invoke/invoke.cc:163]: Codacy Issue found: [Variable 'coeff_max' is assigned a value that is never used.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5469982)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5469982&bid=17926176)

File: src/vt/vrt/collection/balance/lb_invoke/invoke.h [hunk 2]
```diff
@@ -88,14 +84,23 @@ struct LBManager {
...
+    proxy_ = proxy;
+  }
+
+  /*
+   * Tell the manager the LB is finished. This should *not* be called by the
    * user, only by load balancers. Not private/protected as friending every LBs
    * adds too much overhead
    */
-  static void finishedRunningLB(PhaseType phase);
+  void finishedRunningLB(PhaseType phase);
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/lb_invoke/invoke.h:103]: Not salient to this PR, but why not add a cut-out `finished()` method in `BaseLB` to simply call this, and then friend that?

Comment (PhilMiller) [reply src/vt/vrt/collection/balance/lb_invoke/invoke.h:103]: I think that would be making `BaseLB` an 'attorney' and the individual strategies the 'clients'?

**Comment by PhilMiller** [comment]
Fix the build issues, and then this generally looks good.

File: src/vt/vrt/collection/balance/lb_invoke/invoke.h (hunk 2 above)
Comment (lifflander) [reply src/vt/vrt/collection/balance/lb_invoke/invoke.h:104]: That's a good point, should be easily doable.

**Pull request merged**